src/utils/asar.ts

```diff
@@ -0,0 +1,111 @@
+// @ts-ignore
+import asar from 'asar';
+import { toReadableStream } from './buffer';
+import path from 'path';
+import fs from 'fs';
+/*eslint-disable no-unused-vars*/
+import { Readable } from 'stream';
+/*eslint-enable no-unused-vars*/
+
+interface ParsedPath {
+    archive: string;
+    fileName: string;
+}
+
+const ASAR_EXTNAME: string = '.asar';
+
+export default class Asar {
+    private static _instance: Asar = new Asar();
+
+    private _archivePaths: Array<string> = [];
+
+    constructor () {
+        if (Asar._instance)
+            throw new Error('Error: Instantiation failed: Use Asar.getInstance() instead of new.');
+
+        Asar._instance = this;
+    }
+
+    static getInstance () : Asar {
+        return Asar._instance;
+    }
+
+    private _addPath (asarPath: string) {
+        if (this._archivePaths.indexOf(asarPath) === -1)
+            this._archivePaths.push(asarPath);
+    }
+
+    private _findArchivePath (fullPath: string) : string {
+        let currentPath = fullPath;
+        let currentDir  = path.dirname(currentPath);
+
+        while (currentPath !== currentDir) {
+            let isFile = false;
+
+            try {
+                isFile = fs.statSync(currentPath).isFile();
```

Comment:
It will be nice if we can avoid using the synchronous version here.

Reply:
1)
```js
try {
    const stats = await stat(currentPath);

    isFile = stats.isFile();
}
// eslint-disable-next-line no-empty
catch (e) {
}
```
2)
```js
let isFile;

await stat(currentPath).then(() => { isFile = true; }, () => { isFile = false; });
```
3)
```js
private _isFile (fullPath: string) : Promise<boolean> {
    return stat(fullPath)
        .then(() => true, () => false);
}
```
```js
if (await this._isFile(currentPath) && path.extname(currentPath) === ASAR_EXTNAME)
    return currentPath;
```